src/AutoSchedule.cpp

```diff
@@ -99,6 +99,39 @@ struct FStage {
     }
 };
 
+// Check the output_buffers() and propagate any estimates set there into the schedule's
+// estimates. If there are multiple output buffers, any estimates found on them must match
+// or assert-fail will occur. Note that after this function, we can end up in duplicates
+// if both estimate() and set_bounds_estimate() are called for the same arg; that will be
+// caught as an error later on.
```

Comment:
Actually if you have dupicates in estimate(), it won't throw an error. It will use the last defined estimate instead. So the question is, if both estimate() and set_bounds_estimate() are called, which one should we prioritize?

Reply:
Hmm, I has assumed that an error would always result because estimates.size() != out.args().size(), but I suppose you could construct a sequence of calls with a size-match but some estimates unspecified. Yuck.

Re: which should we prioritize... my first assumption was that we should expect exactly one estimate per arg, but I see your point of last-one-wins being useful (presumably this allows us to provide a default and override it elsewhere). With the approach I have here we lose the order of calls unfortunately. Maybe I can restructure this code to do the propagation immediately instead of lazily? Let me take a look.

Reply:
> Maybe I can restructure this code to do the propagation immediately instead of lazily

For the case of multiple output buffers (Func<Tuple>), this is thorny: we probably want to ensure that the estimates match across all of them, if we're doing it immediately, that's not possible. (Although: we could just treat each set_bounds_estimate call for any output buffer as an estimate for all of them, following the same "last one wins" rule...)

Reply:
> Actually if you have dupiclates in estimate(), it won't throw an error. It will use the last defined estimate instead.

I'm not sure this is the case: when I change test/auto_schedule/param.cpp run_test_1() from

    g.estimate(x, 0, 1000).estimate(y, 0, 1000);

to

    g.estimate(x, 0, 1000).estimate(y, 0, 1000).estimate(y, 0, 2000);

I get a test failure:

    Please provide estimate for each dimension of "g"

Reply:
Ah good catch. The region computation code is written in such a way to use the last-defined estimate, but the checker doesn't do that. I will make a PR to fix this bug.